Source/console-checkers/Source/CheckersNotationView.cpp

```diff
@@ -0,0 +1,196 @@
+static constexpr std::string_view s_emptySquareColorHint = "\033[38;5;239m";
+
+// You only see this if there's an error with finding the piece color.
+static constexpr std::string_view s_errorColor = "\033[38;5;206m";
+
+// In Checkers notation, the first numbered position is 1.
+static constexpr int32_t s_positionStart = 1;
+
+static std::string_view GetColorForPiece(const Piece& piece, bool shouldShowHint)
+{
+	static const std::unordered_map<Piece, std::string_view, PieceHash> s_colorMap =
+	{
+		{ { PieceType::Empty, Identity::Neutral }, s_emptySquareColor },
+		{ { PieceType::Pawn, Identity::Red }, s_redPawnColor },
+		{ { PieceType::King, Identity::Red}, s_redKingColor },
+		{ { PieceType::Pawn, Identity::Black }, s_blackPawnColor },
+		{ { PieceType::King, Identity::Black }, s_blackKingColor },
+	};
+
+	static const std::unordered_map<Piece, std::string_view, PieceHash> s_colorHighlightMap =
+	{
+		{ { PieceType::Empty, Identity::Neutral }, s_emptySquareColorHint },
+		{ { PieceType::Pawn, Identity::Red }, s_redPawnColorHint },
+		{ { PieceType::King, Identity::Red}, s_redKingColorHint },
+		{ { PieceType::Pawn, Identity::Black }, s_blackPawnColorHint },
+		{ { PieceType::King, Identity::Black }, s_blackKingColorHint },
+	};
+
+	const auto& relevantColorMap = shouldShowHint ? s_colorHighlightMap : s_colorMap;
+	const auto it = relevantColorMap.find(piece);
+	if (it != relevantColorMap.end())
+	{
+		return it->second;
+	}
+
+	return s_errorColor;
+}
+
+struct CheckersNotationInputSettings {
+
+	struct IndexMapping
+	{
+		// This is the value a player will input.
+		int32_t notationValue = 0;
+
+		// This is the index into the game board that the notation maps to.
+		int32_t index = 0;
+	};
+
+	// An input position token has 2 values. ex. 01 or 23 or 11, etc
+	static constexpr int32_t s_requiredTokenCount = 2;
+
+	// Maps from an input index represented in checkers notation, to a game board index.
+	static constexpr std::array<IndexMapping, 32> s_checkersNotationToIndexMappings =
+	{
+		// Row 1
+		IndexMapping{ 1, 1 },IndexMapping{ 2, 3 },IndexMapping{ 3, 5 }, IndexMapping{ 4, 7 },
+		// Row 2
+		IndexMapping{ 5, 8 },IndexMapping{ 6, 10 },IndexMapping{ 7, 12 }, IndexMapping{ 8, 14 },
+		// Row 3
+		IndexMapping{ 9, 17 },IndexMapping{ 10, 19 },IndexMapping{ 11, 21 }, IndexMapping{ 12, 23 },
+		// Row 4
+		IndexMapping{ 13, 24 },IndexMapping{ 14, 26 },IndexMapping{ 15, 28 }, IndexMapping{ 16, 30 },
+		// Row 5
+		IndexMapping{ 17, 33 },IndexMapping{ 18, 35 },IndexMapping{ 19, 37 }, IndexMapping{ 20, 39 },
+		// Row 6
+		IndexMapping{ 21, 40 },IndexMapping{ 22, 42 },IndexMapping{ 23, 44 }, IndexMapping{ 24, 46},
+		// Row 7
+		IndexMapping{ 25, 49 },IndexMapping{ 26, 51 },IndexMapping{ 27, 53 }, IndexMapping{ 28, 55 },
+		// Row 8
+		IndexMapping{ 29, 56 },IndexMapping{ 30, 58 },IndexMapping{ 31, 60 }, IndexMapping{ 32, 62 },
+	};
+};
+
+CheckersNotationView::CheckersNotationView() = default;
```

Comment:
std::unique_ptr requires it. You get a compile error otherwise. The constructor definition needs to be visible. Pretty sure you ran into this at some point years ago too. Same is true for the destructor.

Reply:
It's the "= default;" that's throwing me off here. I've literally never seen that in a cpp file.